### ISSUE TITLE
FMI directory structure does not support common CPU architectures

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/fmi-standard#302]
**Reported by sjoelund&#x2e;se on 2 Dec 2014 09:40 UTC**
The directory structure of FMI is something like:
```
linux32/modelIdentifier.so
```

But Linux commonly uses different 32-bit (or 64-bit) processors. Like ARM or x86-64. The directory structure needs to be different in order to accommodate all processors.

One possible solution would be to use a name corresponding to OS-CPU. Several alternatives are possible:
```
$ uname -sm | tr "[:upper:] " "[:lower:]-"
linux-x86_64
darwin-x86_64
$ clang -dumpmachine
x86_64-pc-linux-gnu
x86_64-apple-darwin13.0.0
```

----

Migrated-From: https://trac.fmi-standard.org/ticket/302

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 16 Dec 2014 19:48 UTC**
You have a point here. I wonder if such a change can be introduced into the "clarification" release 2.0.1. At least we would need a sound proposal to make sure that everything would be covered after the change. I think that linux-x86-64 as well as specifiers like -pc- or -apple- are too unspecific. Wouldn't we need some version information for the os? For instance:
```
os-version-processor
windows-win32-x86_64
linux-glibc2.19-ix86
linux-glibc2.19-x86_64
darwin-macosx10.10-x86_64
darwin-ios8-ARM
```

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 16 Dec 2014 20:11 UTC**
One might also add a version for windows, in order to enable dynamic linking with a system provided runtime lib, like:
```
windows-msvcr110-ix86
```

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 17 Dec 2014 06:05 UTC**
Some more thoughts on the importance of this issue: 
- "darwin64" as specified in the current FMI 2.0 document, addresses both: MacBook and iPhone, which are not binary compatible. What about WinRT netbooks?
- Static linking, as proposed for the Windows example, is generally bad from a performance point of view. Moreover it causes IT security issues: if a security whole is fixed in a runtime lib, then every statically linked FMU needs to be recreated. This makes binary code export unapplicable from a practical point of view.
- Static linking causes license issues: the LGPL license of libc, for example, gives the right to the user to replace it with a new version (also solving the IT security issue). Vendors had to ship object files or source code and users had to care about security updates for FMUs.

sjoelund: having brought up the issue and certainly invested some more thoughts on it, could you continue working on a proposal for the FMI spec?

**Comment by modelica-trac-importer** [comment]
**Comment by andreas&#46;junghanns on 17 Dec 2014 14:20 UTC**
From the FMI 2.0 text, for everyone as a reference:
```
The FMU must be distributed with at least one implementation, in other words either sources or one of
the binaries for a particular machine. It is also possible to provide the sources and binaries for different
target machines together in one zip file. The following names are standardized: For Windows: “win32”,
“win64”- For Linux: “linux32”, “linux64”. For Mac: “darwin32“, “darwin64” Futhermore, also the names
“VisualStudioX” and “gccX” are standardized and define the compiler with which the binary has been
generated [, for example VisualStudio8]. Further names can be introduced by vendors.
```

While it is correct that the standard currently misses the processor dependency (of course assuming Intel/AMD), it does not restrict the use of additional platform strings that anyone can add.

Because we want to promote interoperability, it would certainly help if accessing the correct platform is not falling back to some heuristic the moment we are outside the standard's definitions.

So we have two issues:
 1. clarify what processor we meant when specifying the existing platforms
 2. defining a clear rule on how to build a platform string, so both FMU exporter and importer will build matching strings, without fracturing the platform space unnecessarily: we want different strings only if the binaries are not compatible! Same string: binaries are compatible!

For 1. a simple clarification is sufficient: Intel/AMD, that is for FMI 2.0.1 (and maybe 1.0.1)

For 2. the os-version-processor seems a good start, but likely not good enough - we need to discuss that and that is more than a clarification and needs to go into a new minor version. That is for FMI 2.1.0. Please provide a good proposal how to build those strings and generate as many examples as you can think of so we can all verify the proposal.

W.r.t. comment 3:
- Given the current rather skewed distribution on the XC repo, we should really worry getting the major platform issues right. I don't see any issues from vendors having problems exchanging their FMUs on WinRT netbooks. Or FMUs on iPhones.
- Dynamic linking is a nice concept but practically so broken that I would not rely on it and the standard was wisely framed the way it is: Yes, you statically link bugs into your FMU, but you tested your FMU and those bugs did not matter, right? The alternative is: you use a new dynamic lib where the old, dormant bug is gone, but new, active bugs are in. And some API changed - and suddenly your stack is broken... and you need to also have all those other dynamically linked libraries that everyone running your FMU has installed. REALLY?!

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 17 Dec 2014 16:03 UTC**
Restriction to x86(_64) appears sensible for now.

I'm not sure regarding static linking though. My PC receives security updates on an almost daily basis, even though all software has been tested by the respective vendors. WinXP is considered a big security risk since Microsoft stopped supporting it with further updates. How can a vendor of an FMU claim that it is safe forever?

By minimizing and clearly documenting the interfaces. An FMU primarily evaluates a function. A safe FMU provided as dynamic link library should make no system calls out of its own, but always go through its dynamically linked environment. Memory allocation and logging are already treated through the calling environment.  It should be easy to document possible further dependencies, in order to simplify deployment despite of dynamic linking and to manage IT security.

**Comment by modelica-trac-importer** [comment]
**Modified by rfranke on 17 Dec 2014 16:41 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 17 Dec 2014 16:41 UTC**
Instead of **static** linking, wouldn't it be possible to package the respective dynamic runtime lib into the binaries/win32 directory?

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 18 Dec 2014 06:35 UTC**
Thinking of a standardized naming scheme that will survive Microsoft's current transition from 32 to 64 bit, the first proposal above appears most appropriate: `uname` is a long standing Posix standard that is available on virtually any platform (see the long table given at http://en.wikipedia.org/wiki/Uname). It exists as command line tool and as C function.

`uname -s` gives the system name, like Linux or Darwin
`uname -r` gives the release of the os
`uname -m` gives the machine architecture

The release number / os version is tricky and requires more discussion, i.e. hardly ready for FMI 2.0.1 -- but could be extended later on. System and machine are quite clear -- see the results for `uname -sm` plus simple char translations above. 

Uname is not available on Windows. It is neither needed there because the system name is constant and machine architecture finds in the environment variable PROCESSOR_ARCHITECTURE -- note that this environment variable automatically adapts if a 32 bit program runs under 64 bit Windows.

Resulting path names for binaries are:
```
windows-x86
windows-amd64
qnx-x86pc
linux-i386
linux-x86_64
linux-armv7l
darwin-i386
darwin-x86_64
...
```

Shouldn't FMI 2.0.1 make that change, i.e. generally standardize on `uname -sm` and specifically define the use of PROCESSOR_ARCHITECTURE under Windows?

**Comment by modelica-trac-importer** [comment]
**Comment by andreas&#x2e;junghanns on 18 Dec 2014 07:19 UTC**
FMI 2.0.1 must be backward compatible and NOT introduce new features: bug fixes (not changing the spec) and clarifications are allowed only.

Your proposal is more than that, so it has to be in 2.1.0.
(Except of course a clarification as to the processor type as stated above.)

Moreover, for the platforms already named, we should try to keep them, if that seems feasible. So at least the ones named already and sufficiently unique to not cause incompatible binaries being listed for the same platform should remain as is.
Otherwise we risk creating confusion for end users of the FMI standard.

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 18 Dec 2014 08:26 UTC**
The new names can also be introduced as extension, meaning that an importer has to support the old and the new name, like win32 and windows-x86. Then it's backwards compatible.

Our current importer already supports dynamic loading of fmi\* names and fmi2\* names. It's not really hard to do such things.

"End users" are generally not stupid. They will understand a change if it makes a lot of sense.

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 18 Dec 2014 16:59 UTC**
Weighting pros and cons, a way forward could be:

Stay with win32 and win64 on Windows platforms, in order to ensure backwards compatibility.

Change the naming and exploit `uname` on Posix systems (e.g. Linux and Darwin).

In section 2.3 of the spec rephrase:

_The following names are standardized: For Windows: “win32”, “win64” - 
For Linux: “linux32”, “linux64”. For Mac: “darwin32“, “darwin64”_

to:

_The following names are standardized: For Windows: “win32”, “win64” assuming x86 processors. On Posix compliant systems the name is formed by concatenating with a dash and only using lower case characters:_
- _the operating system name (uname -s)_
- _the machine hardware identifier (uname -m)_
_Examples are: “linux-i686”, “linux-x86_64” for PC Linux with 32 resp. 64 bit, “linux-armv6l” for Raspberry Pi, “darwin-x86_64” for Mac OS X._

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 19 Dec 2014 04:53 UTC**
An alternative formulation with version information, which requires less discussion on Posix systems:

_The following names are standardized: For Windows: “win32”, “win64” assuming x86 processors. On Posix compliant systems the name is formed by concatenating with a dash and only using lower case characters:
- _the operating system name (uname -s),_
- _the operating system release (uname -r), restricting to the format major.minor.patchlevel_ 
- _the machine hardware identifier (uname -m)_
Examples are: “linux-3.16.0-i686”, “linux-3.16.0-x86_64” for PC Linux with 32 resp. 64 bit, “linux-3.1.9-armv6l” for Raspberry Pi, “darwin-14.0.0-x86_64” for Mac OS X._

**Comment by modelica-trac-importer** [comment]
**Comment by andreas&#46;junghanns on 19 Dec 2014 09:04 UTC**
Currently, a simulation environment will use the platform string to determine, if an FMU will run on the current platform. If the sting is different, it must assume it cannot.

This proposal will fracture the platform space and we should try to minimize that fracturing.
We should at least drop the patchlevel from the OS release. Most minor versions will be forward and backward compatible as well...

While the proposal is simple to implement, it will result in either:
 1. FMUs being unnecessarily rejected as platform incompatible (using simple string equality), or
 2. the simulation environment computing more complex platform compatibility checks.

We have to consider both ways:
 - FMU generated on an old system running in a newer version
 - FMU generated in a new system running in an older
While most systems are largely allowing old executables to run in newer versions, the reverse is often not the case.

Can we find a better way to ensure maximal interoperability?
Better would mean:
 - the standard does not need constant adding of new, compatible platform clusters
 - simple rules for simulation platforms to decide if an FMU generated on platform X will run on current platform Y, rejecting only those it needs to reject

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 19 Dec 2014 09:58 UTC**
The FMI standard should not / hardly can define "compatible platform clusters". This is up to the importing environment. The FMU just provides the version information of the platform it was built on.

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 19 Dec 2014 19:55 UTC**
Here is a new proposal that adds rules for compatibility check and flips the order in the platform string in order to simplify compatibility checks.

_The following names are standardized: For Windows: “win32”, “win64” assuming x86 processors. On Posix compliant systems the name is formed by concatenating with a dash and only using lower case characters:_
- _the machine hardware identifier (uname -m)_
- _the operating system name (uname -s),_
- _the operating system release (uname -r), restricting to the format major.minor.patchlevel_
_Examples are: “i686-linux-3.16.0”, “x86_64-linux-3.16.0” for PC Linux with 32 resp. 64 bit, “armv6l-linux-3.1.9” for Raspberry Pi, “x86_64-darwin-14.0.0” for Mac OS X._

_The importing environment can generally assume that an FMU is running if hardware identifier and os name match and if possible dependencies on dynamic link libraries are satisfied in the target environment [dependecies can be analyzed using tools like dependency walker under Windows or ldd under Linux]. The os release provides further information for exploitation by the importing environment. Moreover it allows to export multiple versions for different os releases._
_

**Comment by modelica-trac-importer** [comment]
**Comment by torstenblochwitz on 10 Feb 2015 13:42 UTC**
Discussion at FMI Meeting:
- The meaning of the existing platform types should be clarified in 2.0.1: Add "assuming x86 processors".

- The sentence "Further names can be introduced by vendors." has to be kept in order to be backwards compatible.

- The above naming convention (except release number) should be added as non-normative text as a guideline for building further names for Posix systems.

- For Windows systems a convention is still missing. The system variable "OS" could be used. But the "Processor_Architecture" variable seems to be not sufficient.

**Comment by modelica-trac-importer** [comment]
**Modified by sjoelund&#46;se on 21 Feb 2015 08:17 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by cxbrooks on 26 Feb 2015 16:05 UTC**
I'm all for allowing different names for the binaries directory, but the old name must be supported or else this is FMI-3.0, not 2.1.  Major version numbers should change when the new system is not compatible with the old system.

A good test is whether fmuCheck-2.0.1 works with the proposed FMU.  If it does, then the proposed FMU is FMU-2.x compliant.  If it does not, then either the FMU needs to be fixed or the FMU version bumped up to FMI-3.x.

This recently came up in OpenModelica, where under Mac OS X, binaries/darwin64 was not being created.  Fortunately, the fix was easy.  Details: https://trac.openmodelica.org/OpenModelica/ticket/3169

**Comment by modelica-trac-importer** [comment]
**Comment by cbertsch on 30 Dec 2017 17:22 UTC**
I like Rüdiger's proposal, 

>The following names are standardized: For Windows: “win32”, “win64” assuming x86 processors. On Posix compliant systems the name is formed by concatenating with a dash and only using lower case characters:
>the machine hardware identifier (uname -m)
>the operating system name (uname -s),
>the operating system release (uname -r), restricting to the format >major.minor.patchlevel
>Examples are: “i686-linux-3.16.0”, “x86_64-linux-3.16.0” for PC Linux with 32 resp. 64 bit, “armv6l-linux-3.1.9” for Raspberry Pi, “x86_64-darwin-14.0.0” for Mac OS X.

but I would change the order to be more similar to the existing names (that should be kept for backwards compatibility!)
* the operating system name (uname -s),
* the machine hardware identifier (uname -m)
* _if necessary:_ the operating system release (uname -r), restricting to the format major.minor.patchlevel)

But there are platforms that contain "-" in the machine hardwares identifier, so we should at least use a different separating character. Underscore is also not possible as separting characters, as there is, e.g., "x86_64"
I have not yet found a rule, which characters are allowed as return values of uname. (http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_utsname.h.html does not specify this)
Looking at https://en.wikipedia.org/wiki/Uname, e.g., a double underscore or "\*" does not seem to be used in the return values of uname and thus could be used as separator
So a directory name could be 

`linux__armv7l__4.9.41` or `linux*armv7l*4.9.41` 

Do we really need the operating system release number in the name or could this information be stored somewhere else? 
* in the modeldescription.xml? --> possibly bad idea, because it is until now platform independent.
* in a separate file in /binaries/platform --> would not allow for the support of different releases within one FMU
* a directory structure within the binaries, i.e. /binaries/linux/armv7l/4.9.41 --> more complicated than existing flat structure

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 30 Dec 2017 17:28 UTC**
> Looking at https://en.wikipedia.org/wiki/Uname, e.g., a double underscore or "\*" does not seem to be used in the return values of uname and thus could be used as separator

`*` Is not valid in filenames on Windows

**Comment by modelica-trac-importer** [comment]
**Comment by andreas&#46;junghanns on 30 Dec 2017 19:50 UTC**
Let's keep in mind that the strings should only be different if the shared or static objects are incompatible. For compatible binaries, the platform string should be the same.
Adding version numbers needlessly, will hinder identifying compatibility.

**Comment by modelica-trac-importer** [comment]
**Comment by cbertsch on 30 Dec 2017 21:57 UTC**
Now that the FMU-SDK 2.0.5 (http://www.qtronic.de/en/fmusdk.html) brings support for Linux and MacOS (Thanks to QTronic!), I have created a fork with experimental support for Raspberry Pi 3 https://github.com/chrbertsch/FMUSDK_fork  - for the time beeing using the name linux-armv7l.

What about `linux__armv7l` as a binaries name?
According to https://raspberrypi.stackexchange.com/questions/45025/raspberry-pi-3-software-compatibility binaries for Raspi 2 and 3 should be compatible (only drivers are incompatible). Thus should we use a common binaries name instead of `linux__armv6l` (Raspi 2) and `linux__armv7l` (Raspi 3)? But `linux__arm` is not ok, since, e.g.,  `linux__armv8l` will be 64bit while the other two are 32bit. So best keep `linux__armv7l` and let the importer know that this could also run on a Raspi 2, right?

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 31 Dec 2017 06:00 UTC**
Why not just use the name of the architecture that the compiler targets; seems simpler than hard-working coding any names. If there is an uncertainty when importing, just try to dlopen multiple shared objects.

**Comment by KarlWernersson** [comment]
Do we need to change this for 2.0.1?
Can it be done in a good forward compatible way?

**Comment by t-sommer** [comment]
Another way would be to adapt the naming scheme for `Bundle-NativeCode` used in [OSGI](https://www.osgi.org/developer/specifications/reference/). They have extensive support for different CPU architectures and OSs. They also define aliases which we could use to assure backwards compatibility.

**Comment by beutlich** [comment]
Cross reference: https://github.com/modelica/ModelicaSpecification/issues/2220

**Comment by xogeny** [comment]
It might be worth having a look at how Rust handles this.  They have what they call a "target triple" (which is, confusing, sometimes a "target quadruple").  The targets they support and how they name them are discussed at https://forge.rust-lang.org/platform-support.html.

Note, they support a LOT of targets.

**Comment by sjoelund** [comment]
The target triple is standard Unix stuff; like:
https://clang.llvm.org/docs/CrossCompilation.html#target-triple

The vendor is almost never needed. And the ABI is usually not needed (except gnueabihf if you require a hardware FPU in your android device). You could have the standard recommend searching for best match (like looking for arm-linux-gnueabihf before arm-linux-gnueabi or arm-linux).

**Comment by beutlich** [comment]
True, for FMI 2.0.1 we can do the minimal clarification change, for FMI 3.0 we can do a proper fix, hopefully.

**Comment by t-sommer** [comment]
Based on the comments above I've created a proposal for a "platform tuple" as a basis for discussions:

The tuple has the general format `<arch>{<arch_sub>}-<sys>{-<abi>{<abi_version>}{<abi_sub>}}`, where:

- `<arch_sub>` = x86_64, i386, i586, i686, arm, thumb, mips

- `<sub>` = for ex. on ARM: v5, v6m, v7a, v7m

- `<sys>` = linux, windows, darwin, netbsd

- `<abi>` = eabi, gnu, android, macho, elf, msvc

- `<abi_version>` = 140 (Visual Studio 14.0)

- `<abi_sub>` = md, mt, mdd, mtd

Examples:

`x86_64-darwin`  -- shared libraries for macOS 
`i586-windows-msvc140mtd` -- static libraries for Visual Studio (with /MTd)
`armv7-linux` -- shared libraries for Raspberry Pi with Ubuntu Linux

**Comment by chrbertsch** [comment]
Can this be closed due to the merged #519?

**Comment by t-sommer** [comment]
Yes. This has been added with #516.
